### PR TITLE
New modules: `ultra/index` and `ultra/align`

### PR DESCRIPTION
**Pull request by sguizard**
Add ultra/index and ulta/align modules. The ultra/pipeline perform both index and align steps. Separate the two steps allow to run indexing once and reuse the index for reads alignment. 

## PR checklist

- [X] This comment contains a description of changes (with reason).
- [X] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [X] If necessary, include test data in your PR.
- [X] Remove all TODO statements.
- [X] Emit the `versions.yml` file.
- [X] Follow the naming conventions.
- [X] Follow the parameters requirements.
- [X] Follow the input/output options guidelines.
- [X] Add a resource `label`
- [X] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [X] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [X] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [X] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/ultra/align/main.nf
Comment (mirpedrol) [inline modules/ultra/align/main.nf:13]: These variables are not used, shouldn't they be provided as the index path?

File: modules/ultra/align/main.nf
Comment (mahesh-panchal) [inline modules/ultra/align/main.nf:16]: Sam files are usually requested to be compressed. Is there a reason not to compress here?

File: modules/ultra/index/main.nf
Comment (mahesh-panchal) [inline modules/ultra/index/main.nf:15]: Files are normally emitted into separate channels. Must these two always be present together?

File: tests/modules/ultra/align/main.nf
Comment (mahesh-panchal) [inline tests/modules/ultra/align/main.nf:25]: The SAMTOOLS_SORT process ( albeit quick ) is unnecessary.

File: modules/ultra/align/main.nf
Comment (sguizard) [reply modules/ultra/align/main.nf:16]: uLTRA produce sam files as output, qnd the container do not contains samtools for compressing.

File: modules/ultra/index/main.nf
Comment (sguizard) [reply modules/ultra/index/main.nf:15]: To be usable, the pickle files and the db files have to be both together for aligning with uLTRA align. I thought it would be the best way to keep them together.

File: tests/modules/ultra/align/main.nf
Comment (sguizard) [reply tests/modules/ultra/align/main.nf:25]: Agree, I clean this.

File: modules/ultra/align/main.nf
Comment (mahesh-panchal) [reply modules/ultra/align/main.nf:16]: Usually nf-core requests a mulled container be built in these cases such as for the modules like Chromap , bwa mem2 , etc

Comment (mahesh-panchal) [reply modules/ultra/align/main.nf:16]: https://nf-co.re/developers/modules#software-requirements

File: modules/ultra/index/main.nf
Comment (mahesh-panchal) [reply modules/ultra/index/main.nf:15]: That's fair. I think it's not a good idea to split files that must be used together.

File: modules/ultra/align/main.nf
Comment (sguizard) [reply modules/ultra/align/main.nf:16]: Please tells me if I am wrong, mulled containers are interesting if you can pipe the result directly into samtools to avoid intermediary files? uLTRA can't print output on STDOUT. So if I mulled it with samtools, compress the sam file, at the end I will both the sam file and the bam file, which is less space efficient than the sam file alone.

Comment (mahesh-panchal) [reply modules/ultra/align/main.nf:16]: You can always `rm` the sam file. It's not the most ideal situation if you're not using `process.scratch` since your work directory will ultimately need space for both, but by the end of the process you'll only have the bam. 

Depending on the tool, process substitution can also be an option to let you pipe directly to another process. 
```bash
command --outfile >( samtools sort ... - ) ...
```

Comment (sguizard) [reply modules/ultra/align/main.nf:16]: Thanks for the resources, I had a look to the mulled containers and how to apply it to the module. 
For the uLTRA module, the management of the outputs is made with two parameters, the `--prefix` option and the output directory parameter. Therefore, I can't use process substitution to avoid temporary files.
The only option for compressing the sam file into bam will require space for both files. 
If you think that have a bam file as output is a mandatory requirement, I prepared a mulled container and I can update the code.

Comment (mahesh-panchal) [reply modules/ultra/align/main.nf:16]: I'll let the core team have the final decision on that. 
@mirpedrol What's your take?

Comment (mirpedrol) [reply modules/ultra/align/main.nf:16]: I would usually say to compress all files, but if this requires even more memory it doesn't have much sense right?

Comment (mahesh-panchal) [reply modules/ultra/align/main.nf:16]: The issue isn't memory but temporary disk space. While the process is running, it will need space for both the sam and the bam until until `rm *.sam` is called at the end of the process. I don't think it's much of an issue, and it's generally better practice to use `process.scratch`, but not everyone has access to scratch space.

Comment (mirpedrol) [reply modules/ultra/align/main.nf:16]: I see, then converting from sam to bam seems a good option 👍

Comment (sguizard) [reply modules/ultra/align/main.nf]: I added `samtools sort` to sort and convert the output `sam` in the same time. 
The module now fully comply to the output requirement.

File: modules/ultra/index/main.nf
Comment (mahesh-panchal) [inline modules/ultra/index/main.nf:22]: Need `args2` now with samtools present.

**Review by mahesh-panchal — APPROVED** [review]
Aside from `args2` missing for samtools, all good. Thank you.

**Comment by sguizard** [comment]
Thanks for the reviews!

**Pull request merged**
